src/Core/PostgreSQL/ConnectionHolder.h

```diff
@@ -20,11 +21,20 @@ class ConnectionHolder
 {
 
 public:
-    ConnectionHolder(PoolPtr pool_, ConnectionPtr connection_) : pool(pool_), connection(std::move(connection_)) {}
+    ConnectionHolder(PoolPtr pool_, ConnectionPtr connection_, bool auto_close_)
+        : pool(pool_)
+        , connection(std::move(connection_))
+        , auto_close(auto_close_)
+    {}
 
     ConnectionHolder(const ConnectionHolder & other) = delete;
 
-    ~ConnectionHolder() { pool->returnObject(std::move(connection)); }
+    ~ConnectionHolder()
+    {
+        if (auto_close)
+            connection.reset();
+        pool->returnObject(std::move(connection));
```

Comment:
It'd be more generic to have a policy for returning an object to the pool. But it'll make the change more complicated and make the PR scope wider

Reply:
what do you mean by "a policy for returning an object to the pool"? for now the rule is -- it has to be always returned, otherwise the pool gets broken.